File: src/main/java/util/days/AbstractDay.java
package util.days;

import java.io.IOException;
import java.time.Duration;
import java.time.Instant;

public abstract class AbstractDay {

    private final int year;
    private final int day;

    public AbstractDay(int year, int day) {
        this.year = year;
        this.day = day;
    }

    public abstract String part1(String input);

    public abstract String part2(String input);

    private String loadInput() {
        try {
            return InputFetcher.fetchInput(year, day);
        } catch (IOException | InterruptedException | UnsuccessfulDownloadException e) {
            throw new RuntimeException(e);
        }
    }

    public void main() {
        System.out.println(year + " day " + day + ":\n");
        String input = this.loadInput();

        Instant start = Instant.now();
        String result = part1(input);
        Instant stop = Instant.now();
        printResult(1, result, Duration.between(start, stop));

        start = Instant.now();
        result = part2(input);
        stop = Instant.now();
        printResult(2, result, Duration.between(start, stop));
    }

    private void printResult(int part, String result, Duration duration) {
        System.out.println("part " + part + ": " + result);
        System.out.println(DynamicDuration.formatDuration(duration));
    }
}

File: src/main/java/aoc/AbstractDay.kt
package aoc

import util.days.InputFetcher
import java.time.Duration
import java.time.Instant

abstract class AbstractDay(private val year: Int, private val day: Int) {

    abstract fun part1(input: String): Any
    abstract fun part2(input: String): Any

    fun executeParts() {
        val input = InputFetcher.fetchInput(year, day)
        println("$year Day $day:\n")

        executePart(1) { part1(input) }
        executePart(2) { part2(input) }
    }

    private fun executePart(partNumber: Int, part: () -> Any) {
        val result = executeAndMeasure { part() }
        println("Part $partNumber: ${result.first}")
        println(" ${formatDuration(result.second)}")
    }

    private fun executeAndMeasure(part: () -> Any): Pair<Any, Duration> {
        val start = Instant.now()
        val result: Any = part()
        val end = Instant.now()

        return Pair(result, Duration.between(start, end))
    }
}